### ISSUE TITLE
Rust+SDL2+GL: OpenGL function invocations fails with "exception thrown: RuntimeError: invalid index into function table"

### ISSUE THREAD

**Issue by therocode** [emscripten-core/emscripten#11715]
I am successfully compiling my Rust project for the wasm32-unknown-emscripten target. I'm using rust-sdl2 and gl_generator with a 4.3 core GL profile.

When built for native targets, it works and my GL commands produce stuff on the screen as expected. However, the wasm32 version errors with the following:
![image](https://user-images.githubusercontent.com/3073690/88366427-b9eb3780-cdbb-11ea-9752-8e401c6cbf1f.png)

**gl functions fail**
In this example, glEnable is what is failing, but I've tried putting other gl functions before it, and they all fail. glScissor fails with the same `invalid index into function table`, glClearColor however fails with `Exception thrown: RuntimeError: function signature mismatch,RuntimeError: function signature mismatch`. Aside from the differing error, the callstack is the same.

**getprocaddress**
I tried with: SDL2's function loading, and with emscripten_GetProcAddress. The error stayed the same for both. I also tried without executing any GetProcAddress loading code at all and the error also did not change. Code that I used:
Emscripten version:
```
let gl = habanero::gl::Gl::load_with(|s| {
            dbg!("in load", s);
            let c_string = std::ffi::CString::new(s).unwrap();
            let res = unsafe {
                emscripten::emscripten::emscripten_GetProcAddress(c_string.as_ptr()) as *const _
            };
            dbg!(res);
            res
        });
```
SDL2 version:
```
let gl = habanero::gl::Gl::load_with(|s| window.video.gl_get_proc_address(s) as *const _);
```

**Details about my setup**
I'm using emscripten `1.39.20` and bulding with `cargo build --target=wasm32-unknown-emscripten` with this cargo config:
```
[target.wasm32-unknown-emscripten]
rustflags = [
    "-C", "link-arg=-s", "-C", "link-arg=USE_SDL=2",
    "-C", "link-arg=-s", "-C", "link-arg=FULL_ES3=1",
]
```
I'm on Windows 10 if that matters.
cargo --version gives `cargo 1.46.0-nightly (4f74d9b2a 2020-07-08)`

For SDL2 I'm using rust-sdl2 version "0.34". 
I'm using the Struct Generator function of gl_generator, detailed here: https://crates.io/crates/gl_generator
Edit: I have now also tried with the Global Generator, which made no difference.

**Comment by kripken** [comment]
Does using `SDL_GL_GetProcAddress` work? (that should call `emscripten_GetProcAddress`, so probably it won't make a difference, but that is what we test in our test suite)

If not, it should be straightforward to debug this, as GetProcAddress should return an integer which we can verify is the location of `glEnable` in the table. And that integer should be used unmodified later in the indirect call. Might be easiest to debug in a build with something like `-O2 --profiling -s WASM=0` (in the wasm2js output it's easy to inspect the "wasm" and add debug logging; if you can provide such a build I can take a look).

Also maybe @tlively has ideas for debugging.

**Comment by therocode** [comment]
Thanks, some extra info since:

- I tried both `SDL_GL_GetProcAddress` and `emscripten_GetProcAddress` and they indeed produce the same pointer values.
- I got a bit further past the GL calls failing by using the StaticGenerator GL loader which in fact doesn't load anything at all, but uses the gl functions as statically available by the platform. This fails on desktop for obvious reasons, but works in the emscripten build. Doesn't seem ideal though when the GetProcAddress based solutions should be working.
- When getting past the GL problem using the above method, I'm still getting errors on the SDL2 commands instead:
![image](https://user-images.githubusercontent.com/3073690/88617518-6d5c7080-d0c9-11ea-98c6-0b8bc0449425.png)
Which makes me think that maybe there's a wider problem that manifests both in the gl and sdl2 calls somehow? Not sure.

I'll do some further debugging with the given flags and also I'll try to setup a minimal repro and provide source+build, cheers.

**Comment by therocode** [comment]
When trying a build with WASM=0, I'm getting
`note: emcc: error: wasm2js does not support source maps yet (debug in wasm for now)` which I'm not really sure what it means.

**Comment by therocode** [comment]
Progress: Of course when I do a minimal repro case it actually works... I'm kinda expecting something to do with memory corruption due to how the more complex case is setup when it comes to the [emscripten_set_main_loop ](https://gist.github.com/therocode/08b2dba27313119677d3e779b9d8ca2d) function and rust lifetimes of objects... I'll investigate more and report back

**Comment by therocode** [comment]
Alright, I've fixed the problem and it was indeed due to lifetime issues. Turns out that the rust code floating around to setup the emscripten main loop doesn't at all account for the lifetime of your closure: https://gist.github.com/therocode/08b2dba27313119677d3e779b9d8ca2d

The closure given to set_main_loop_callback is not kept alive in any way, so it along with any data it was holding is goin to go out of scope. I guess the solution is to change this code to actually stash the closure and keep it alive in a thread_local storage until emscripten_cancel_main_loop is called, at which point ownership of it would be returned to the caller.

So, in the end it turned out it's more a rust-side error than emscripten. I'll for sure write up an article how to set things up properly. Feel free to close this issue if you have nothing to add, thanks!
